src/ynEIGEN/TokenStakingNode.sol

```diff
@@ -108,6 +118,129 @@ contract TokenStakingNode is
+        uint32 _startBlock,
+        uint256 _shares,
+        IStrategy _strategy,
+        uint256[] memory _middlewareTimesIndexes,
+        bool updateTokenStakingNodesBalances
+    ) public onlyTokenStakingNodesWithdrawer {
+
+        IDelegationManager _delegationManager = tokenStakingNodesManager.delegationManager();
+
+        IDelegationManager.Withdrawal[] memory _withdrawals = new IDelegationManager.Withdrawal[](1);
+        {
+            IStrategy[] memory _strategiesArray = new IStrategy[](1);
+            _strategiesArray[0] = _strategy;
+            uint256[] memory _sharesArray = new uint256[](1);
+            _sharesArray[0] = _shares;
+            _withdrawals[0] = IDelegationManager.Withdrawal({
+                staker: address(this),
+                delegatedTo: _delegationManager.delegatedTo(address(this)),
+                withdrawer: address(this),
+                nonce: _nonce,
+                startBlock: _startBlock,
+                strategies: _strategiesArray,
+                shares: _sharesArray
+            });
+        }
+
+        IERC20 _token = _strategy.underlyingToken();
+        uint256 _balanceBefore = _token.balanceOf(address(this));
+
+        {
+            bool[] memory _receiveAsTokens = new bool[](1);
+            _receiveAsTokens[0] = true;
+            IERC20[][] memory _tokens = new IERC20[][](1);
+            _tokens[0] = new IERC20[](1);
+            _tokens[0][0] = _token;
+
+            _delegationManager.completeQueuedWithdrawals(
+                _withdrawals,
+                _tokens,
+                _middlewareTimesIndexes,
+                _receiveAsTokens
+            );
+        }
+
+        uint256 _actualAmountOut = _token.balanceOf(address(this)) - _balanceBefore;
+        IWrapper _wrapper = IYieldNestStrategyManager(tokenStakingNodesManager.yieldNestStrategyManager()).wrapper();
+        IERC20(_token).forceApprove(address(_wrapper), _actualAmountOut); // NOTE: approving also token that will not be transferred
+        (_actualAmountOut, _token) = _wrapper.wrap(_actualAmountOut, _token);
+
+        queuedShares[_strategy] -= _shares;
+        withdrawn[_token] += _actualAmountOut;
+
+        if (updateTokenStakingNodesBalances) {
+            // Actual balance changes only if slashing occured. choose to update here
+            // only if the off-chain considers it necessary to save gas
+            IYieldNestStrategyManager(tokenStakingNodesManager.yieldNestStrategyManager()).updateTokenStakingNodesBalances(
+                _token
+            );
+        }
+
+        emit CompletedQueuedWithdrawals(_shares, _actualAmountOut, address(_strategy));
+    }
+
+    /**
+     * @notice Deallocates tokens from the withdrawn balance and approves them for transfer.
+     * @param _token The ERC20 token to deallocate.
+     * @param _amount The amount of tokens to deallocate.
+     */
+    function deallocateTokens(IERC20 _token, uint256 _amount) external onlyYieldNestStrategyManager {
```

Comment:
should this also transfer tokens out?

approve is a idempotent operation while 

`withdrawn[_token] -= _amount;`

is not.

The way this function is structured is strange in that way in that it decreases the internal amount but tokens are not guaranteed to move.

Reply:
https://github.com/yieldnest/yieldnest-protocol/pull/164/commits/7412dffe7b9378ee45baec6f68bcf37a5b123994